### PR TITLE
Allow customizing request delay interval and maximum number of retries.

### PR DESCRIPTION
**Pull request by nvthongswansea**
Allow customizing request delay interval and maximum number of retries. Since there are some users applying moderate/large amount of concurrency on terraform, customizing these parameters would solve some issues relating to the rate limit. Changes:
- Read custom request delay interval from the arg `request_delay_int`. If `request_delay_int` is omitted, read from the env var `GRIDSCALE_TF_REQUEST_DELAY_INT`.
- Read maximum number of retries from the arg `max_n_retries`. If `max_n_retries` is omitted, read from the env var `GRIDSCALE_TF_MAX_N_RETRIES`.
- Update the docs.

### CODE REVIEW

File: website/docs/index.html.markdown [hunk 1]
```diff
@@ -53,5 +53,7 @@ The following arguments are supported:
...
 * `token` - (Optional) This is an API-Token for the gridscale API. It can be created [in the panel](https://my.gridscale.io/APIs/). The created token needs to have full access to be usable by Terraform. If omitted, the GRIDSCALE_TOKEN environment variable is used.
-* `api_url` - (Optional) The URL for the API. By default this is set to "https://api.gridscale.io". Do not add a "/" character at the end.
+* `api_url` - (Optional) The URL for the API. If omitted, the GRIDSCALE_URL environment variable may be used. By default this is set to "https://api.gridscale.io". Do not add a "/" character at the end.
 * `http_headers` - (Optional) Custom HTTP headers sent to gridscale API. If omitted, the GRIDSCALE_TF_HEADERS environment variable may be used.
+* `request_delay_int` - (Optional) Custom request delay interval in Ms. This time interval is used to delay the synchronous request checks, or delay retryable requests. If omitted, the GRIDSCALE_TF_REQUEST_DELAY_INT environment variable may be used. By default this is set to 1000.
```
Comment (bkircher) [inline website/docs/index.html.markdown]: Change `Ms` to `ms` here.

File: gridscale/provider.go [hunk 2]
```diff
@@ -41,6 +41,18 @@ func Provider() terraform.ResourceProvider {
 				DefaultFunc: schema.EnvDefaultFunc("GRIDSCALE_TF_HEADERS", nil),
 				Description: "Custom HTTP headers",
 			},
+			"request_delay_int": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("GRIDSCALE_TF_REQUEST_DELAY_INT", nil),
+				Description: "Custom request delay interval in Ms. This time interval is used to delay the synchronous request checks, or delay retryable requests.",
```
Comment (bkircher) [inline gridscale/provider.go]: Change `request_delay_int` to `request_delay_interval` and `GRIDSCALE_TF_REQUEST_DELAY_INT` to `GRIDSCALE_TF_REQUEST_DELAY_INTERVAL`, respectively.

**Review by bkircher — CHANGES_REQUESTED** [review]
Okay in general. I would suggest a name change, though, for that delay interval setting.

The main thing: Do we really need `max_n_retries` to be 5? Why is not 1 the default?

I would expect the provider to fail more often, but these are then 
1. Probably real bugs that will fail again in the next retry anyway
2. Are not bugs in the backend but other failures that should not be retried anyway (examples is permission denied)

`max_n_retries` of 1 would, I suppose, decrease load on the system overall.

What is your opinion?

**Comment by nvthongswansea** [comment]
@bkircher Answers:
1. The default value 5 of `max_n_retries` is from `gsclient-go` (we've used this default value since the beginning). This `max_n_retries` should be used in case user want to deal with the rate limiter when using moderate amount of concurrency. Usually, these 2 variables are not used in normal case in which everything runs ok.
2. Which requests are retriable, which are not? These 2 questions' answers belong to `gsclient-go`.

**Comment by bkircher** [comment]
>   The default value 5 of max_n_retries is from gsclient-go (we've used this default value since the beginning). This max_n_retries should be used in case user want to deal with the rate limiter when using moderate amount of concurrency. Usually, these 2 variables are not used in normal case in which everything runs ok.

I know what it does. I just think the default value should be 1 instead of 5. (Example: In gscloud, the value is set to 1.)


>    Which requests are retriable, which are not? These 2 questions' answers belong to gsclient-go.

Not all HTTP methods make sense to retry. Also, retrying 429 is different from say 503. One might include a `Retry-After:` header the other might not.

Other responses, say 500, 404, or 400 do not make sense to retry at all. Is gsclient-go taking care of this?

**Comment by nvthongswansea** [comment]
@bkircher Again, the retriable request concern belongs to `gsclient-go`, precisely, [here](https://github.com/gridscale/gsclient-go/blob/master/request.go#L279).

**Comment by bkircher** [comment]
> Again, the retriable request concern belongs to gsclient-go, precisely, here.

Thanks for the link. I think this is a completely wrong handling of retries.

429 should re-tried indefinitely (or stopped on signals like SIGINT)

503 should also be retried but blocking on `Retry-After`value. There a `max_n_retries` value might make sense.

As for 
- 409
- 424 and
- 500

Those should not be retried at all. IMO.

**Comment by nvthongswansea** [comment]
> > Again, the retriable request concern belongs to gsclient-go, precisely, here.
> 
> Thanks for the link. I think this is a completely wrong handling of retries.
> 
> 429 should re-tried indefinitely (or stopped on signals like SIGINT)
> 
> 503 should also be retried but blocking on `Retry-After`value. There a `max_n_retries` value might make sense.
> 
> As for
> 
> * 409
> * 424 and
> * 500
> 
> Those should not be retried at all. IMO.

Could you open an issue in `gsclient-go` repo, please? I will set the default n retries to 1 in this PR for our tf provider.

**Comment by bkircher** [comment]
> Could you open an issue in gsclient-go repo, please? 

Okay, will do.

> I will set the default n retries to 1 in this PR for our tf provider.

If I understand the code in gsclient-go correctly now, we want something like > 1 (e.g. the 5 you suggested). But that doesn't make it correct. It just might work in more cases by default than giving 1 here.

**Comment by bkircher** [comment]
So IMO, we can merge this here. It is just not the correct solution (or full solution) to the problem at hand. @nvthongswansea  Okay with merging here?

**Comment by nvthongswansea** [comment]
> If I understand the code in gsclient-go correctly now, we want something like > 1 (e.g. the 5 you suggested). But that doesn't make it correct. It just might work in more cases by default than giving 1 here.

The default `max_n_retries` in `gsclient-go` doesn't relate to the `max_n_retries` in terraform provider, since we use `gsclient.NewConfiguration()` instead of `gsclient.DefaultConfiguration()` to initialize a client in terraform. So that in terraform provider, we have to manually set the default value of `max_n_retries` ([here](https://github.com/gridscale/terraform-provider-gridscale/blob/master/gridscale/config.go#L26)).

**Comment by nvthongswansea** [comment]
> So IMO, we can merge this here. It is just not the correct solution (or full solution) to the problem at hand. @nvthongswansea Okay with merging here?

Yep. But let me change the default value of `max_n_retries` in our tf provider firstly.

**Comment by nvthongswansea** [comment]
@bkircher done.

**Pull request merged**